zio-crypto/shared/src/main/scala/zio/crypto/mac/HMAC.scala

```diff
@@ -0,0 +1,163 @@
+package zio.crypto.mac
+
+import zio._
+import zio.crypto.ByteHelpers
+
+import java.nio.charset.Charset
+import java.security.MessageDigest
+import javax.crypto.spec.SecretKeySpec
+import javax.crypto.{ KeyGenerator, Mac, SecretKey }
+
+sealed class HMACAlgorithm(val name: String)
+
+object HMACAlgorithm {
+  case object HMACSHA1   extends HMACAlgorithm("HmacSHA1")
+  case object HMACSHA256 extends HMACAlgorithm("HmacSHA256")
+  case object HMACSHA384 extends HMACAlgorithm("HmacSHA384")
+  case object HMACSHA512 extends HMACAlgorithm("HmacSHA512")
+}
+
+case class HMACSerializedKey(value: String)     extends AnyVal
+case class HMACSecretKey(underlying: SecretKey) extends AnyVal
+case class HMACObject[Kind](value: Kind)        extends AnyVal
+
+object HMAC {
+
+  type HMAC = Has[HMAC.Service]
+
+  trait Service {
+    def sign(m: Array[Byte], k: HMACSecretKey): Task[HMACObject[Array[Byte]]]
```

Comment:
Pure functions? Also Chunk, etc.

Reply:
`java.crypto.Mac.init`, which is used in the body of `sign` throws if the key is invalid. `Chunk` addressed here and elsewhere :+1: